### PR TITLE
Start adding a performance section to the guide.

### PR DESCRIPTION
**Pull request by adamreichold**
We discussed this quite often and I though that even if it is only a single thing, just making a start would improve our chances of getting it finished eventually.

### CODE REVIEW

File: guide/src/performance.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
+# Performance
+
+To achieve the best possible performance, it is useful to be aware of several tricks and sharp edges concerning PyO3's API.
+
+## `extract` versus `downcast`
+
+Pythonic API implemented using PyO3 are often explicitly polymorphic, i.e. they will accept `&PyAny` and try to turn this into multiple more concrete types to which the requested operation is applied. This often leads to chains of calls to `extract`, e.g.
```
Comment (davidhewitt) [inline guide/src/performance.md:7]:
```suggestion
Pythonic API implemented using PyO3 are often polymorphic, i.e. they will accept `&PyAny` and try to turn this into multiple more concrete types to which the requested operation is applied. This often leads to chains of calls to `extract`, e.g.
```

**Review by davidhewitt — APPROVED** [review]
Nice, thanks for kicking this off!

One day it would be nice to remove this performance pitfall. Something to investigate after removing the pool 😄

**Comment by adamreichold** [comment]
@davidhewitt I added another subsection on implicit access to GIL token. Please have another look before this is merged.

Also do you remember any other topics when wanted to discuss here? I do think I remember that there about four different things we wanted to have here, but I am unable find a list anywhere in the older issues, pull requests or discussions.

**Comment by adamreichold** [comment]
> Also do you remember any other topics when wanted to discuss here?

Is dictionary dispatch, i.e. look-up tables based on type object identity, something we want to discuss here or is this too specialized?

**Comment by davidhewitt** [comment]
Personally I think that's quite application-specific, but I'm not against having that section if you want to write it.

Some other topics I recalled over the day:
- Overhead of conversions (i.e. `PyList` vs `Vec` etc.) 
- Threading / allow_threads
- String `intern!`
- `Vec<u8>` becoming `list[int]` (and thus being very slow)
- `#[pyo3(get)]` deep-cloning non-Py data

We can potentially move these into an issue rather than write them all in this PR?

**Comment by adamreichold** [comment]
> Threading / allow_threads

Shouldn't this be covered by the "parallelism" section?

> We can potentially move these into an issue rather than write them all in this PR?

Yes, I'd like to limit this PR to the two small sections already written.

> Personally I think that's quite application-specific, but I'm not against having that section if you want to write it.

Ok, let's include a note in the issue. Certainly not in this PR.

**Comment by davidhewitt** [comment]
Sorry for the slow reply - I got a bit sidetracked by the various PRs I have pushed in the last 24h...

> Shouldn't this be covered by the "parallelism" section?

Yes, though I think it can't hurt to call it out i.e. "how do I multithread? See parallelism section". More visibility :)

**Pull request merged**